File: Stack/minStack.cpp
#include <bits/stdc++.h>
using namespace std;

class minStack
{
private:
    stack<pair<int, int>> stack;
    int minElement;

public:
    minStack()
    {
    }

    void push(int val)
    {
        if (stack.empty())
        {
            stack.push({val, val});
        }

        else
        {
            if (val < stack.top().second)
            {
                stack.push({val, val});
            }
            else
            {
                minElement = stack.top().second;
                stack.push({val, minElement});
            }
        }
    }

    void pop()
    {
        stack.pop();
    }

    int top()
    {
        return stack.top().first;
    }

    int getMin()
    {
        return stack.top().second;
    }
};

File: Stack/minStack.py
class MinStack:
    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))

        else:
            if val < self.stack[-1][1]:
                self.stack.append((val, val))
            else:
                minElement = self.stack[-1][1]
                self.stack.append((val, minElement))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]
